### PR TITLE
 Remove decorator laziness, support tables reuse, nested Serial

### PR DESCRIPTION
**Pull request by guillaumegenthial**
# Description

Remove decorator laziness, support tables reuse, nested Serial.

BREAKING CHANGES

## Type of change

- [x] Feature
- [ ] Fix
- [ ] Release

## Checklist

- [x] I have followed the [CONTRIBUTING](https://github.com/criteo/deepr/blob/master/docs/CONTRIBUTING.rst) guidelines.
- [x] I have updated the [CHANGELOG](https://github.com/criteo/deepr/blob/master/docs/CHANGELOG.rst).

### CODE REVIEW

File: deepr/layers/base.py [hunk 1]
```diff
@@ -336,103 +343,57 @@ def __init__(self, offset, n_in=1, n_out=1, inputs=None, outputs=None, name=None
...
-            def _forward(self, tensors, mode: str = None):
-                return fn(*self._args, **self._kwargs).forward(tensors, mode)
-
+        # Check legacy use case
+        if "tensors" not in parameters:
+            msg = f"'tensors' should be the first parameter of {fn.__name__}. "
+            msg += "Using a decorator on a Layer constructor has been deprecated. "
+            msg += "The decorator will have no impact (please remove it)."
+            LOGGER.error(msg)
+            return fn
```
Comment (guillaumegenthial) [inline deepr/layers/base.py:361]: Decorator works on the same cases as before, but on the constructor case, is not applied.
We can either keep it that way, or just fail (would prefer, that would force people to remove the useless decorators).

File: deepr/prepros/base.py [hunk 2]
```diff
@@ -68,25 +71,28 @@ def __repr__(self) -> str:
 
     def __call__(self, dataset: tf.data.Dataset, mode: str = None) -> tf.data.Dataset:
         """Alias for apply"""
-        return self.apply(dataset, mode=mode)
+        if TableContext.is_active():
+            return self.apply(dataset, mode=mode)
```
Comment (guillaumegenthial) [inline deepr/prepros/base.py:75]: wrap apply in a table context so that it's transparent for the user.

File: deepr/prepros/base.py [hunk 3]
```diff
@@ -103,91 +109,48 @@ def __init__(self, offset)
...
-            """Create prepro during first call (lazy)"""
-            return fn(*self._args, **self._kwargs).apply(dataset=dataset, mode=mode)
-
+    # Check legacy use case
+    if "dataset" not in parameters:
+        msg = f"'dataset' should be the first parameter of {fn.__name__}. "
+        msg += "Using a decorator on a Prepro constructor has been deprecated. "
+        msg += "The decorator will have no impact (please remove it)."
+        LOGGER.error(msg)
+        return fn
```
Comment (guillaumegenthial) [inline deepr/prepros/base.py:128]: same change for the prepro decorator

File: deepr/prepros/combinators.py [hunk 4]
```diff
@@ -58,27 +58,35 @@ def __init__(
...
-            _fuse(*to_flat_tuple(preprocessors), num_parallel_calls=num_parallel_calls)
-            if fuse
-            else to_flat_tuple(preprocessors)
-        )
+        self.preprocessors = to_flat_tuple(preprocessors)
         self.fuse = fuse
         self.num_parallel_calls = num_parallel_calls
 
+        # Iterable of preprocessors used by the apply method
+        self._preprocessors = (
```
Comment (guillaumegenthial) [inline deepr/prepros/combinators.py:66]: keep both the original list of preprocessors and the fused version for nested unrolling

File: deepr/prepros/combinators.py [hunk 5]
```diff
@@ -58,27 +58,35 @@ def __init__(
...
         return dataset
 
 
 def _fuse(*preprocessors: Prepro, num_parallel_calls: int = None) -> Tuple[Prepro, ...]:
     """Group Map and Filter in _FusedMap and _FusedFilter"""
 
+    def _flatten(prepros):
+        for prepro in prepros:
+            if isinstance(prepro, Serial):
+                yield from _flatten(prepro.preprocessors)
```
Comment (guillaumegenthial) [inline deepr/prepros/combinators.py:86]: unroll nested Serial

File: deepr/config/base.py [hunk 6]
```diff
@@ -15,9 +15,16 @@
 
 KEY_POSITIONAL_ARG = "*"
 
-EVAL_MODE_INSTANCE = "instance"
+EVAL_MODE_INSTANCE = "as_instance"  # cls(**params)
 
-EVAL_MODE_SKIP = "skip"
+EVAL_MODE_CONSTRUCTOR = "as_constructor"  # partial(cls, **params)
+
+EVAL_MODE_DICT = "as_dict"  # {"type": cls, **params}
```
Comment (guillaumegenthial) [inline deepr/config/base.py:22]: we need to find good names for those 3 modes

File: docs/CHANGELOG.rst [hunk 7]
```diff
@@ -16,15 +16,31 @@ Doctest run in CI.
 Changed
 ~~~~~~~
 copy_dir job will now overwrite the target by default
+- Nested support for ``prepros.Serial``
```
Comment (rom1504) [inline docs/CHANGELOG.rst:19]: should we consider these changes as breaking ?

**Comment by rom1504** [comment]
not passing the new doctest tests ;)

File: docs/CHANGELOG.rst (hunk 7 above)
Comment (guillaumegenthial) [reply docs/CHANGELOG.rst:19]: indeed!

**Pull request merged**